### ISSUE TITLE
consider not including img alt text as part of surrounding text properties

### ISSUE THREAD

**Issue by aaronpk** [microformats/microformats2-parsing#16]
This became more critical once I started working on [removing images from posts containing a photo](https://github.com/aaronpk/XRay/issues/52) in XRay. 

Given a post like [tantek's](http://tantek.com/2018/003/t3/first-npsf-earlygang-year), which starts with two imgs inside the e-content both containing alt text, the parser results in the `name` and `content` properties both beginning with the img alt text. When XRay goes to remove the img tags from the HTML, it should have also removed the img's alt text from the plaintext versions, but it can't since the mf2 parser has already included it. 

Aside from that case, I have also heard several anecdotal cases where the alt text doesn't produce a good result for plaintext values.

The proposal is to consider not including img alt text as part of surrounding text properties. It would only be included for implied values and when there is a mf2 class on the img tag itself.

**Comment by aaronpk** [comment]
There is a case where it makes sense to leave the alt text as part of the text content, for example 

```
<div class="e-content"><img src="fancy-A.png" alt="A">aron</div>
```

should be parsed as

```
  "content": {
    "html": "<img src=\"fancy-A.png\" alt=\"A\">aron",
    "value": "Aaron"
  }
```

so this is a case where the alt text should be left in. 

I think the right rule is that if the parser has pulled out the alt text into a property (#2) then it should be removed from the text content. That would allow a consumer to put the post back together if it's consuming the alt text.

**Comment by gRegorLove** [comment]
I'm watching the [2020 microformats session](https://indieweb.org/2020/Pop-ups/Microformats) and I think I like the proposed rule. I have [examples](https://gregorlove.com/2022/04/pandemic-chic/) similar to Tantek's. 

Current parse (truncated):

```json
"properties": {
    "photo": [
        {
            "value": "https://gregorlove.com/site/assets/files/6050/pxl_20210410_185435944.1000x0-is.jpg",
            "alt": "selfie wearing a black face mask and red sunglasses with flower decorations while sitting under an orange umbrella"
        }
    ],
    "content": [
        {
            "html": "<figure class=\"no-margin\"><img alt=\"selfie wearing a black face mask and red sunglasses with flower decorations while sitting under an orange umbrella\" class=\"u-photo wide-bleed-image\" src=\"https://gregorlove.com/site/assets/files/6050/pxl_20210410_185435944.1000x0-is.jpg\">\n<figcaption>\n<p>Pandemic chic from April 2021. Sunglasses borrowed from <a class=\"h-card\" href=\"https://www.instagram.com/lizlemonskinnypuppy/\">Laurie</a>.</p>\n</figcaption>\n</figure>",
            "value": "selfie wearing a black face mask and red sunglasses with flower decorations while sitting under an orange umbrella\nPandemic chic from April 2021. Sunglasses borrowed from Laurie.",
            "lang": "en"
        }
    ],
}
```

So in general, it would make sense to me if the parser removed that matching alt text and the result was:

```json
"properties": {
    "photo": [
        {
            "value": "https://gregorlove.com/site/assets/files/6050/pxl_20210410_185435944.1000x0-is.jpg",
            "alt": "selfie wearing a black face mask and red sunglasses with flower decorations while sitting under an orange umbrella"
        }
    ],
    "content": [
        {
            "html": "<figure class=\"no-margin\"><img alt=\"selfie wearing a black face mask and red sunglasses with flower decorations while sitting under an orange umbrella\" class=\"u-photo wide-bleed-image\" src=\"https://gregorlove.com/site/assets/files/6050/pxl_20210410_185435944.1000x0-is.jpg\">\n<figcaption>\n<p>Pandemic chic from April 2021. Sunglasses borrowed from <a class=\"h-card\" href=\"https://www.instagram.com/lizlemonskinnypuppy/\">Laurie</a>.</p>\n</figcaption>\n</figure>",
            "value": "Pandemic chic from April 2021. Sunglasses borrowed from Laurie.",
            "lang": "en"
        }
    ],
}
```

I don't know the best spec language for this. Really rough attempt at an update for [parsing an e-](https://microformats.org/wiki/microformats2-parsing#parsing_an_e-_property), prefixed lines with [new] and [adjusted]:

> parsing an e- property
> * `html` ... [no change]
> * `value`: the `textContent` of the element after
>   * dropping any nested <script> & <style> elements;
>   * [new] removing any nested <img> elements that have a `u-*` microformat property
>   * [adjusted] replacing any other nested `<img>` elements with their alt attribute, if present; otherwise their src attribute, if present, adding a space at the beginning and end, resolving the URL if it’s relative;

We would probably need to add some text explaining why some images are removed and others replaced.

**Comment by gRegorLove** [comment]
I think the same rule would work for [parsing a p-](https://microformats.org/wiki/microformats2-parsing#parsing_a_p-_property). 

With Tantek's example linked above:

```json
"properties": {
    "name": [
        "a jpg. a jpg. \ud83c\udf06 Made it to the first #NPSF #earlygang of the year, did in-betweeners abs, and 6:30 workout with a brutal burnout that was really its own workout. But wow pretty sunrise. Plus 50+ deg F? I\u2019ll take it. #100PDPD\n\n#justshowup #darknesstodawn #wakeupthesun #fromwhereirun #NovemberProject #sunrise #latergram #nofilter"
    ],
    "photo": [
        {
            "value": "https://igx.4sqi.net/img/general/width960/476_g7yruXflacsGr7PyVmECefyTBMB_R99zmPQxW7pftzA.jpg",
            "alt": "a jpg. "
        },
        {
            "value": "https://igx.4sqi.net/img/general/width960/476_zM3UgU9JHNhom907Ac_1WCEcUhGOJZaNWGlRmev86YA.jpg",
            "alt": "a jpg. "
        }
    ],
}
```

It would become:

```json
"properties": {
    "name": [
        "\ud83c\udf06 Made it to the first #NPSF #earlygang of the year, did in-betweeners abs, and 6:30 workout with a brutal burnout that was really its own workout. But wow pretty sunrise. Plus 50+ deg F? I\u2019ll take it. #100PDPD\n\n#justshowup #darknesstodawn #wakeupthesun #fromwhereirun #NovemberProject #sunrise #latergram #nofilter"
    ],
    "photo": [
        {
            "value": "https://igx.4sqi.net/img/general/width960/476_g7yruXflacsGr7PyVmECefyTBMB_R99zmPQxW7pftzA.jpg",
            "alt": "a jpg. "
        },
        {
            "value": "https://igx.4sqi.net/img/general/width960/476_zM3UgU9JHNhom907Ac_1WCEcUhGOJZaNWGlRmev86YA.jpg",
            "alt": "a jpg. "
        }
    ],
}
```

**Comment by gRegorLove** [comment]
And here is an [example ](https://aaronparecki.com/2020/12/22/14/oauth) of `img` with `alt`, no `u-*` property, inside `e-content`:

```json
"properties": {
    "content": [
        {
            "html": "<p>If you've been putting off setting aside some time to learn about OAuth, now is your chance!</p>\n\n<p>Over the last year, I've helped thousands of developers learn about OAuth by hosting live workshops in person, online events through O'Reilly and Okta, as well as by publishing videos on YouTube! I'm super thrilled to announce that I just finished packaging up the workshop and have launched it as a new course, \"<a href=\"http://oauth.wtf/course\">The Nuts and Bolts of OAuth 2.0</a>\"!</p>\n\n<p>The course is 3.5 hours of video content, quizzes, as well as interactive exercises with a guided learning tool to get you quickly up to speed on OAuth, OpenID Connect, PKCE, best practices, and tips for protecting APIs with OAuth.</p>\n\n<p>The course is <a href=\"http://oauth.wtf/course\">available now on Udemy</a>, and if your company has a Udemy for Business subscription you can find it there as well! If you download the app, you can even sync the video course to a mobile device to watch everything offline!</p>\n\n<p><a href=\"http://oauth.wtf/course\"><img src=\"https://aaronparecki.com/2020/12/22/14/image-1.jpg\" alt=\"Screenshot of OAuth course on Udemy\"></a></p>\n\n<p>The exercises in the course will walk you through the various OAuth flows to set up an OAuth server, get an access token, use a refresh token, and learn the user's name and email with OpenID Connect. You can see a sneak peek of the tool that interactively helps you debug your apps at <a href=\"https://oauth.school/\">oauth.school</a>.</p>\n\n<h2>Free OAuth Videos</h2>\n\n<p>I've also got a bunch of videos about OAuth available on YouTube for you to watch at any time! Take a look at my <a href=\"https://oauth.net/videos/\">curated playlist of videos</a> where you'll find everything from <a href=\"https://www.youtube.com/watch?v=td00HpMI3Q8\">live sketch notes</a> of a conversation about PKCE and OAuth security, to a description of <a href=\"https://www.youtube.com/watch?v=espX8qKjywI\">OAuth phishing</a>, to&#xA0;details on <a href=\"https://www.youtube.com/watch?v=CHzERullHe8\">why you shouldn't use the implicit flow</a>.</p>",
            "value": "If you've been putting off setting aside some time to learn about OAuth, now is your chance!\nOver the last year, I've helped thousands of developers learn about OAuth by hosting live workshops in person, online events through O'Reilly and Okta, as well as by publishing videos on YouTube! I'm super thrilled to announce that I just finished packaging up the workshop and have launched it as a new course, \"The Nuts and Bolts of OAuth 2.0\"!\nThe course is 3.5 hours of video content, quizzes, as well as interactive exercises with a guided learning tool to get you quickly up to speed on OAuth, OpenID Connect, PKCE, best practices, and tips for protecting APIs with OAuth.\nThe course is available now on Udemy, and if your company has a Udemy for Business subscription you can find it there as well! If you download the app, you can even sync the video course to a mobile device to watch everything offline!\nScreenshot of OAuth course on Udemy\nThe exercises in the course will walk you through the various OAuth flows to set up an OAuth server, get an access token, use a refresh token, and learn the user's name and email with OpenID Connect. You can see a sneak peek of the tool that interactively helps you debug your apps at oauth.school. Free OAuth Videos\nI've also got a bunch of videos about OAuth available on YouTube for you to watch at any time! Take a look at my curated playlist of videos where you'll find everything from live sketch notes of a conversation about PKCE and OAuth security, to a description of OAuth phishing, to\u00a0details on why you shouldn't use the implicit flow."
        }
    ]
}
```

The alt text "Screenshot of OAuth course on Udemy" is in the `content.value` and would continue to parse that way with this spec update, as expected.

I'll update this comment with more examples as I find them. I'm still looking for an example of of `img` with `alt`, no `u-*` property, inside `p-name`:
